types/tipset.go

```diff
@@ -2,155 +2,148 @@ package types
+			}
+			if blk.ParentWeight != weight {
+				return NoTipSet, errors.Errorf("Inconsistent block parent weights %d and %d", weight, blk.ParentWeight)
+			}
 		}
-	}
-	return ts, nil
-}
-
-// AddBlock adds the provided block to this tipset.
-// PRECONDITION: this block has the same height parent set as other members of ts.
-func (ts TipSet) AddBlock(b *Block) error {
-	if len(ts) == 0 {
-		id := b.Cid()
-		ts[id] = b
-		return nil
+		// Reject duplicate blocks (by CID).
+		c := blk.Cid()
+		if cids[c] {
+			return NoTipSet, errors.Errorf("Duplicate block CID %s", c)
+		}
+		cids[c] = true
+		sorted[i] = blk
 	}
 
-	h, err := ts.Height()
-	if err != nil {
-		return err
-	}
-	p, err := ts.Parents()
-	if err != nil {
-		return err
-	}
-	weight, err := ts.ParentWeight()
-	if err != nil {
-		return err
-	}
-	if uint64(b.Height) != h {
-		return errors.Errorf("block height %d doesn't match existing tipset height %d", uint64(b.Height), h)
-	}
-	if !b.Parents.Equals(p) {
-		return errors.Errorf("block parents %s don't match tipset parents %s", b.Parents.String(), p.String())
-	}
-	if uint64(b.ParentWeight) != weight {
-		return errors.Errorf("bBlock parent weight: %d doesn't match existing tipset parent weight: %d", uint64(b.ParentWeight), weight)
-	}
+	// Sort blocks by ticket.
+	sort.Slice(sorted, func(i, j int) bool {
+		cmp := bytes.Compare(sorted[i].Ticket, sorted[j].Ticket)
+		if cmp == 0 {
+			// Break ticket ties with the block CIDs, which are distinct.
+			cmp = bytes.Compare(sorted[i].Cid().Bytes(), sorted[j].Cid().Bytes())
+		}
+		return cmp < 0
+	})
 
-	id := b.Cid()
-	ts[id] = b
-	return nil
+	return TipSet{sorted}, nil
 }
 
-// Clone returns a shallow copy of the TipSet.
-func (ts TipSet) Clone() TipSet {
-	r := TipSet{}
-	for k, v := range ts {
-		r[k] = v
-	}
+// Defined checks whether the tipset is defined.
+// Invoking any other methods on an undefined tipset will result in undefined behaviour (c.f. cid.Undef)
+func (ts TipSet) Defined() bool {
+	return len(ts.blocks) > 0
+}
 
-	return r
+// Len returns the number of blocks in the tipset.
+func (ts TipSet) Len() int {
+	return len(ts.blocks)
 }
 
-// String returns a formatted string of the TipSet:
-// { <cid1> <cid2> <cid3> }
-func (ts TipSet) String() string {
-	return ts.ToSortedCidSet().String()
+// At returns the i'th block in the tipset.
+// An index outside the half-open range [0, Len()) will panic.
+func (ts TipSet) At(i int) *Block {
+	return ts.blocks[i]
 }
 
-// Equals returns true if the tipset contains the same blocks as another set.
-// Equality is not tested deeply.  If blocks of two tipsets are stored at
-// different memory addresses but have the same cids the tipsets will be equal.
-func (ts TipSet) Equals(ts2 TipSet) bool {
-	return ts.ToSortedCidSet().Equals(ts2.ToSortedCidSet())
+// IsSolo tests whether the tipset has a single block.
+func (ts TipSet) IsSolo() bool {
```

Comment:
This is, at best, unnecessary. `ts.Len()==1` is exactly the same length as `ts.IsSolo()`, but the meaning of the former is unambiguous. It would take me a few guesses to get to IsSolo means contains one block.

Reply:
I am a little shocked at how often we do this check (this is a smell in itself). The problem is we seem to be doing it for different reasons in each case. So in that sense, this function obscures rather than adds semantic meaning to its code.

Categorizing usages:
* 2 are related to whether the tipset is a genesis block. This logic could be encapsulated within expected.
* 2 are optimizations that check to see if we can avoid reprocessing messages when the tipset contains only one block. I'd like to see this fix so all cases are optimized.
* The rest are used in tests, and are actually asserting the tipset has the right number of blocks, not that the tipset is in a special state (e.g. a state of Soloness)

Reply:
Fair enough, reverted. I don't want to try to apply other optimizations or refactors here.